CIP-????/README.md

```diff
@@ -0,0 +1,476 @@
+Plutus Core is a deliberately conservative language: for the most part it is just the untyped lambda calculus.
+The proposed new features for sums-of-products are also conservative, and follow a very typical pattern.
+But they are not _quite_ as standard as the untyped lambda calculus.
+
+Moreover, a change like this would be very painful to reverse, so we should seriously consider the costs before proceeding.
+
+### Discussion
+
+#### Why is case not strict?
+
+This makes it different to all other Plutus Core terms, but we believe it is the expected behaviour: _no_ language has strict case destructuring.
+
+Consider e.g.
+
+```haskell
+case (Just 1) of 
+   Just x -> print "Success!"
+   Nothing -> error "Failure!"
+```
+
+Even in a strict language like Rust, nobody would expect this to evaluate the `Nothing` branch and evaluate to an error.
+
+Furthermore, this additional laziness has significant performance benefits, see Appendix 2 for more discussion.
+
+#### Is this still worth it if users aren't decoding the script context?
+
+The issues with decoding the script context only apply to users who actually try and decode it; many people have moved on to just using the underlying Data object instead.
+Arguably that saps the motivation for this proposal.
+
+There are two responses to this:
+1. This proposal seems worth it on the basis of its other benefits (benefits 2 and 3).
+2. Removing the problems with decoding the script context may allow users to return to that approach, which can be more ergonomic (and potentially cheaper, since builtin calls are expensive).
+
+### Alternatives
+
+#### 1. Use Sums _and_ Products Instead
+
+The current solution is a sums-of-products solution, i.e. we have an introduction and elimination form for objects that have _both_ a tag and a list of fields.
+We could instead separate these two parts out and have an introduction and elimination form for sums, likewise for products.
+
+That would look something like this:
+```
+-- Tags a term with a tag
+(tag i t)
+-- Inspects a tagged term for the tag, and passes the 
+-- inner term to the case function corresponding to the tag
+(case t cs)
+
+-- Constructs a product with the given fields
+(prod ts)
+-- Accesses the i'th field of the given product
+(index i t)
+```
+
+This is cleaner in some ways, and was the first prototype we implemented, but it has the following problems:
+- It adds two more term constructors to the language
+- It is significantly slower in practice, because the combined operation of processing a sum-of-products is extremely common
+
+#### 2. Bind Case Variables in the Syntax for Case Branches
+
+The design presented here allows the case branches to be arbitrary terms, which must be evaluated and then applied to the fields.
+
+A more complex design would be to have the bindings for the variables be part of the case expression instead. That is, the current design does this:
+
+```
+(case (constr 1 a b) (\x -> t1) (\x y -> t2))
```

Comment:
Here it seems that indexing the constructors start from 0.

Reply:
Yes, I realised it was inconsistent and I'm going to redo everything starting from 0.

Reply:
Should be fixed.